src/binary-reader.cc

```diff
@@ -2678,6 +2677,7 @@ Result BinaryReader::ReadElemSection(Offset section_size) {
     }
 
     // For backwards compat we support not declaring the element kind.
+    Type elem_type = Type::Void;
```

Comment:
Huh, I'm surprised we need to parse these as Type::Void element segments that later get auto-resolved (in the BinaryReaderIr and BinaryReaderInterp and SharedValidator) by looking at the table type. Is there a test case (or other case involving pre-reference-types modules) that requires this...? I don't quite understand the situation where this is needed.


From the current spec it seems like there are exactly 8 options, and every choice either implicitly indicates funcref (0 and 4), explicitly encodes elemkind which must be funcref (1 through 3), or explicitly encodes reftype (5-7). When is it implicit?

Reply:
I'm not exactly sure what the history is here, but there seems to be circumstances under which elem_type is not read from the binaryen.  Currently we only read and element type `if (flags & (SegPassive | SegExplicitIndex)) {`.

There are certainly tests that fail if we assume FuncRef like the old code did... but I don't have a precise description to hand.

Reply:
Specifically `elem.wast` in the spec tests fails if we default to FuncRef here... investigating more.

Reply:
I think the issue here may be in the writer side, specifically wast2json (and ultimately ir.cc ElemSegment::GetFlags). Any externref-typed segment needs to have SegUseElemExprs **and** SegExplicitIndex both set, or else the type is implicitly funcref. (Maybe we should rename `SegUseElemExprs` to `SegUseElemTypeAndExprs` to be clearer...)

Reply:
You were right.. much nicer fix now..